docs/src/architecture/backend_architecture.md

```diff
@@ -0,0 +1,220 @@
+      |> put_in([:player_timestamps, player_id], timestamp)
+
+    {:noreply, state}
+  end
+```
+
+Every action handler updates the game state accordingly. We're managing game
+updates through a tick rate. The runner sends a message to itself every 30ms and
+does a broadcast of the new state. Currently, this is being done on the handler
+that matches with `:update_state`.
+
+You'll notice the `init` function on the `PlayWebSocket` process does (among other things) the following:
+
+```elixir
+:ok = Phoenix.PubSub.subscribe(DarkWorldsServer.PubSub, "game_play_#{game_id}")
+```
+
+This allows the socket processes to receive state updates they can then relay to the player.
+
+Below are two diagrams summarizing the whole flow.
+
+### New Game
+
+```mermaid
+sequenceDiagram
+    participant Player (Unity Client)
+    participant Server
+    participant Engine
+    participant Runner
+    participant PlayerWebSocket
+    Player (Unity Client)->>Server: Start Game
+    Server->>Engine: Start Runner
+    Engine->>Runner: Spawn child
+    Runner->>Engine: Runner PID
+    Engine->>Server: PID
+    Server->>Player (Unity Client): game_session_id
+    Player (Unity Client)->>Server: ws://.../play/:game_session_id
+    Server->>PlayerWebSocket: Handle connection
+    PlayerWebSocket->>PlayerWebSocket: PubSub.subscribe( game_session_topic)
+```
+
+### Gameplay
+
+```mermaid
+sequenceDiagram
+    participant Player (Unity Client)
+    participant PlayerWebSocket
+    participant Runner
+    Player (Unity Client)->>PlayerWebSocket: {"action": "move", "value": "right"}
+    PlayerWebSocket->>Runner: cast(runner_pid, {:play, player_id, {:move, :right}})
+    Runner->>Runner: Move player (update internal state)
+    Runner->>PlayerWebSocket: Broadcast(game_session_topic, {:game_update, new_state})
+    PlayerWebSocket->>Player (Unity Client): new_state
+```
+
+# Comments/Omissions
+
+## State management and Rust NIFs
+
+In the architecture walkthrough above, we glossed over how the game state is handled. Currently, it consists of two things:
+
+- A list of players, each of which has an `id`, a position, and a health value.
+- A `board`, which is just a matrix indicating what each cell on the map contains (whether it's empty, has a player in it, etc).
+
+This state is kept inside the `Runner` process, but *state transitions* (players moving, attacking, etc) are computed in Rust. To call Rust code from Elixir we use [Rustler](https://github.com/rusterlium/rustler), which allows us to write `NIF`s; a way to call low level performant code inside the Erlang VM. You can read more about them [here](https://www.erlang.org/doc/tutorial/nif.html) and [here](https://www.erlang.org/doc/man/erl_nif.html).
+
+All the Rust game state code is located inside the `native/gamestate` directory. The functions exposed to Elixir are all in the `lib.rs` file. Here's the function that we call to move players around the map:
+
+```rust
+#[rustler::nif(schedule = "DirtyCpu")]
```

Comment:
We're not using DirtyCpu anymore (perhaps we can reference the load test report in which we decided to remove it)